src/stream/src/executor/source/source_executor.rs

```diff
@@ -245,27 +245,57 @@ impl<S: StateStore> SourceExecutor<S> {
             .stack_trace("source_recv_first_barrier")
             .await
             .unwrap();
+        let epoch = barrier.epoch.prev;
 
         if let Some(mutation) = barrier.mutation.as_ref() {
             if let Mutation::Add { splits, .. } = mutation.as_ref() {
                 if let Some(splits) = splits.get(&self.ctx.id) {
                     self.stream_source_splits = splits.clone();
+                    for split_impl in &mut self.stream_source_splits {
+                        if let Some(recover_state) = self
+                            .split_state_store
+                            .try_recover_from_state_store(split_impl.id(), epoch)
+                            .await?
+                        {
+                            *split_impl = recover_state;
+                        }
+                    }
+
+                    // save source_id -> [split_id] to state store
+                    self.split_state_store
+                        .save_split_assignment(splits, epoch)
+                        .await?;
                 }
             }
+        } else {
+            self.stream_source_splits =
+                match self.split_state_store.load_split_assignment(epoch).await? {
+                    None => vec![],
+                    Some(assignments) => {
+                        let mut mock_state = Vec::with_capacity(assignments.len());
```

Comment:
What does `mock` mean here?

Reply:
I tried to mock a `SplitImpl` with just `id` to reuse some code. Will find a better naming

Reply:
I guess that we should handle Add and Update here, and just leave unreachable for other branches.